workflow/Snakefile

```diff
@@ -0,0 +1,229 @@
+import yaml 
+import json
+import os
+import pandas as pd
+
+configfile: "config/config.yaml"
+
+## load biosamples config file
+biosamples_config_file = config["biosamplesTable"]
+biosamples_config = pd.read_table(biosamples_config_file, na_values="").fillna("None").set_index("biosample", drop=False)
+
+## get list of biosamples
+all_biosamples = list(biosamples_config.index.values)
+
+## get TSS and genefile names for each biosample 
+TSS_files = []
+gene_files = []
+for x in all_biosamples:
+	tss_file = config['params_candidate']['genome_tss']
+	gene_file = config['params_neighborhoods']['genes']
+	if biosamples_config.loc[x, "TSS"] != "None":
+		tss_file = biosamples_config.loc[x, 'TSS']
+	if biosamples_config.loc[x, "genes"] != "None":
+		gene_file = biosamples_config.loc[x, 'genes']
+	TSS_files.append(tss_file)
+	gene_files.append(gene_file)
+				
+biosamples_config["TSS"] = TSS_files
+biosamples_config["genes"] = gene_files
+
+
+RESULTS_DIR = config['predictions_results_dir']
+
+def get_accessibility_file(wildcards):
+	default_accessibility_feature = biosamples_config.loc[wildcards.biosample, 'default_accessibility_feature']
+	return biosamples_config.loc[wildcards.biosample, default_accessibility_feature]
+
+## overall rule to run pipeline
+rule all:
+	input:
+		narrowPeaks_sorted = expand(os.path.join(RESULTS_DIR, "{biosample}", "Peaks", "macs2_peaks.narrowPeak.sorted"), biosample=all_biosamples),
+		candidateRegions = expand(os.path.join(RESULTS_DIR, "{biosample}", "Peaks", "macs2_peaks.narrowPeak.sorted.candidateRegions.bed"),
+	    biosample=all_biosamples),
+		enhancerLists = expand(os.path.join(RESULTS_DIR, "{biosample}", "Neighborhoods", "EnhancerList.txt"),  biosample=all_biosamples),
+		allPutative = expand(os.path.join(RESULTS_DIR, "{biosample}", "Predictions", "EnhancerPredictionsAllPutative.txt.gz"), biosample=all_biosamples),
+		enhPredictions = expand(os.path.join(RESULTS_DIR, "{biosample}", "Predictions", "EnhancerPredictions.csv"), biosample=all_biosamples),
+		allPred = os.path.join(RESULTS_DIR,"AllPredictions.txt.gz"),
+		altTSS = TSS_files,
+		metricsDirectory = expand(os.path.join(RESULTS_DIR, "{biosample}", "Metrics", "QCSummary.p"), biosample=all_biosamples)
+
+	
+## call macs2 -- if multiple accessibility inputs for one biosample, will aggregate into one output
+rule call_macs_peaks: 
+	input:
+		accessibility = get_accessibility_file
+	params:
+		pval = config['params_macs']['pval'],
+		out_dir = config["predictions_results_dir"]
+	conda:
+		"envs/abcenv.yml"
+	output: 
+		narrowPeak = os.path.join(RESULTS_DIR, "{biosample}", "Peaks", "macs2_peaks.narrowPeak")
+	shell: 
+		""" 
+		macs2 callpeak -f AUTO -g hs -p {params.pval} -n macs2 --call-summits --outdir {params.out_dir}/{wildcards.biosample}/Peaks -t {input.accessibility}
```

Comment:
"-g hs" is a parameter setting to say we're running with the human genome.  This is fine for now but ultimately should be a parameter that is changed if user is using mouse genome